src/main/scala/com/microsoft/hyperspace/Hyperspace.scala

```diff
@@ -165,19 +171,31 @@ class Hyperspace(spark: SparkSession) {
   }
 }
 
-object Hyperspace {
-  private lazy val context = new ThreadLocal[HyperspaceContext]
+object Hyperspace extends Logging {
+  // Maintain HyperspaceContext for each Spark session, since Hyperspace depends on
+  // the session's properties such as configs, etc.
+  // Value as (thread id, hyperspace context) to enforce one client thread per Spark session
+  // as Hyperspace doesn't guarantee thread safety.
+  private[hyperspace] lazy val contexts =
+    new ConcurrentHashMap[SparkSession, (Long, HyperspaceContext)]().asScala
 
   private[hyperspace] def getContext(spark: SparkSession): HyperspaceContext = {
-    val ctx = context.get()
-    if (ctx == null || !ctx.spark.equals(spark)) {
-      // Ensure the Spark session the caller provides is the same as
-      // the one HyperspaceContext is using because Hyperspace depends on the
-      // session's properties such as configs, etc.
-      context.set(new HyperspaceContext(spark))
+    val ctx = contexts.get(spark)
+    if (!ctx.isDefined) {
+      contexts.synchronized {
+        if (contexts.size >= HyperspaceConf.maxSupportedSparkSessions(spark)) {
+          throw HyperspaceException("Too many Spark sessions uses Hyperspace." +
+            "Please increase the config: spark.hyperspace.maxSessions")
+        }
+        contexts.put(spark, (Thread.currentThread().getId, new HyperspaceContext(spark)))
+      }
+    } else if (ctx.get._1 != Thread.currentThread().getId) {
+      logWarning(s"Hyperspace is not thread safe for threads using one Spark session. " +
+        "Please be aware of it. Current thread id: ${Thread.currentThread().getId}, " +
+        s"Previous thread id: ${ctx.get._1}")
+      contexts.put(spark, (Thread.currentThread().getId, ctx.get._2))
     }
-
-    context.get()
+    contexts.get(spark).get._2
```

Comment:
Instead of looking up again, you can do the following?
```scala
    val ctx = contexts.get(spark)
    if (!ctx.isDefined) {
      contexts.synchronized {
        if (contexts.size >= HyperspaceConf.maxSupportedSparkSessions(spark)) {
          throw HyperspaceException("")
        }
        val newCtxt = new HyperspaceContext(spark)
        contexts.put(spark, (Thread.currentThread().getId, newCtxt))
        newCtxt
      }
    } else {
      if (ctx.get._1 != Thread.currentThread().getId) {
        logWarning(s"")
        contexts.put(spark, (Thread.currentThread().getId, ctx.get._2))
      }
      ctx.get._2
    }
```